### PR TITLE
Don't throw unobserved exception from AdaptedPipeline.ReadInputAsync

### PR DESCRIPTION
**Pull request by halter73**
When I was investigating #1249 using Kestrel's SampleApp I saw the following unobserved exceptions.

```
   {"The decryption operation failed, see inner exception."}
   {"The certificate chain was issued by an authority that is not trusted"}
   at System.Net.Security._SslStream.EndRead(IAsyncResult asyncResult)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncTrimPromise`1.Complete(TInstance thisRef, Func`3 endMethod, IAsyncResult asyncResult, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.AspNetCore.Server.Kestrel.Adapter.Internal.LoggingStream.<ReadAsync>d__17.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.AspNetCore.Server.Kestrel.Adapter.Internal.AdaptedPipeline.<ReadInputAsync>d__9.MoveNext()
   
TaskCanceledException "The request was aborted"
      at Microsoft.AspNetCore.Server.Kestrel.Internal.Http.SocketInput.CheckConnectionError()
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Http.SocketInput.GetResult()
   at Microsoft.AspNetCore.Server.Kestrel.Internal.Http.SocketInputExtensions.<ReadAsyncAwaited>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.AspNetCore.Server.Kestrel.Adapter.Internal.LoggingStream.<ReadAsync>d__17.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Microsoft.AspNetCore.Server.Kestrel.Adapter.Internal.AdaptedPipeline.<ReadInputAsync>d__9.MoveNext()
```

When I removed the connection filters, I still saw "The request was aborted" TaskCanceledExceptions. Those aren't fixed by this PR. I think it is better to wait until #1277 until fixing the non ConnectionAdapter related unobserved exceptions.

### CODE REVIEW

File: test/Microsoft.AspNetCore.Server.KestrelTests/ConnectionAdapterTests.cs [hunk 1]
```diff
@@ -109,6 +108,46 @@ public class ConnectionAdapterTests
             }
         }
 
+        [Fact]
+        public async Task ThrowingOnReadDoesNotUnobservedException()
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/ConnectionAdapterTests.cs]: nit: DoesNot**Throw**

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 2]
```diff
@@ -404,10 +404,8 @@ public void Start()
         /// </summary>
         public Task StopAsync()
         {
-            if (!_requestProcessingStopping)
-            {
-                _requestProcessingStopping = true;
-            }
+            _requestProcessingStopping = true;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Why this change?

File: test/Microsoft.AspNetCore.Server.KestrelTests/ConnectionAdapterTests.cs [hunk 3]
```diff
@@ -109,6 +108,54 @@ public class ConnectionAdapterTests
...
+            var unobservedExceptionThrown = false;
+
+            EventHandler<UnobservedTaskExceptionEventArgs> unobservedExceptionHandler = (sender, args) =>
+            {
+                unobservedExceptionThrown = true;
+            };
+
+            try
+            {
+                TaskScheduler.UnobservedTaskException += unobservedExceptionHandler;
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.Server.KestrelTests/ConnectionAdapterTests.cs]: This just looks like trouble waiting to happen. You may end up observing unobserved tasks in tests since they run in parallel.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs (hunk 2 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: The check is completely redundant.

File: test/Microsoft.AspNetCore.Server.KestrelTests/ConnectionAdapterTests.cs (hunk 3 above)
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.KestrelTests/ConnectionAdapterTests.cs]: True, but after the PR there shouldn't be any 😄

Obviously there's a risk that other test code could do this though. Would you recommend reporting exception details so we could track down any unobserved tasks this happens to catch? Only fail if it's the specific exception instance we threw? Or just remove the test?

Comment (davidfowl) [inline test/Microsoft.AspNetCore.Server.KestrelTests/ConnectionAdapterTests.cs]: Remove the test and instead just have the handler in the sample. It can be used as a quick sanity check if we need to verify behavior. We could also potentially move it to an end to end test instead (part of the server tests).

Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/ConnectionAdapterTests.cs]: Create a separate test project just for this 💩 :trollface:
